docs/references/architecture/adr-113-mempool-lanes.md

```diff
@@ -0,0 +1,312 @@
+gossip protocol, p2p channel, p2p bandwidth capacity, etc.
+For an MVP, probably some (most?) of these components can be common to all lanes.
+In further versions, we can see which of these need to be per-lane.
+What is clear is that each lane will need its own `CList` data structure, right from the beginning.
+* `CList` **need to be different**
+  * if using a common `CList`, the broadcast routine becomes overly complicated,
+    whereas each lane having its own `CList`, the broadcast routine is very simple
+    (advance the highest priority, non-empty lane)
+* cache **can be shared** in MVP
+  * lock contention: we think it won't be a big deal performance-wise for an MVP
+  * cache eviction of entries: lanes with a lot of traffic may force eviction on TXs
+    on lanes with little traffic.
+    * For an SDK application this is irrelevant: SDK implements replay protection (nonce)
+  * it actually doesn't make sense to split the cache per lane, and it would complicate the code significantly.
+* p2p channel **need to be different** in MVP for different lanes
+  * p2p channels already have priorities implemented.
+  * we can implement priorities of lanes by assigning them to different p2p channels
+    as a first approach
+  * currently mempool uses one channel: we would reserve a _channel range_ for the lanes
+* p2p bandwidth capacity **can be shared** in MVP
+  * bandwidth capacity per lane is nice to have, but if the MVP doesn't implement it,
+    it's still very usable
+* gossip protocol: each lane could select its own gossip protocol.
+  For the future, when we have more than one gossip protocol to choose from.
+* lane/mempool capacity (# of txs, total # of bytes, bytes per tx) **can be different** easily,
+  with little extra complexity on config.
+  * bytes per tx: ok to share one value for MVP... not so sure
+  * \# of txs: probably not all use cases properly addressed if we share a common limit
+  * total # of bytes: if the other two are per-lane, this can just be a "generous" value
+    (and we'd fine tune the other two)
+  * IDEA: Leave this for the end: MVP is viable without it, but use cases less customizable
+* re-check **can be shared** in MVP
+  * with `PrepareProposal`, it becomes _always_ mandatory, so no point in making it per-lane
+
+ADR: This is good info, should probably go at the beginning,
+"These are our decisions on what to implement and what not to.
+The rest of the document only tackles the items reported as "to implement" here.
+
+### Lane coordination
+
+There should be some kind of Lane Coordinator that receives uncategorised txs and assigns them a lane.
+This could be implemented in the reactor or inside `CListMempool`.
+* For an MVP, probably it's easier to put the coordination logic in `CListMempool`.
+* There will be a _native_ or _default_ lane, similar to the _native VLAN_ in an 802.1Q network.
+* `CheckTx` will (optionally) return lane information, and only then it can be put in a lane.
```

Comment:
So how can we know when a lane is used to broadcast txs from other lanes?

Reply:
With the design we're currently considering (the one sketched in these notes) you would detect a cheating peer because:
* you receive a TX on the p2p channel corresponding to lane A
* you then checktx the TX and your app tells you the TX belongs to lane B (lane A being "more privileged" than B) 

@hvanz Something we haven't discussed, and that @cason's question makes me think of. We will need some property that requires the app at _all nodes_ to provide consistent lane information. Otherwise we might end up in the same unfortunate situation we have now, where the outcome of `CheckTx` might depend on some `app.toml` config (e.g., min fees) and therefore `CheckTx` is not consistent across nodes.